### ISSUE TITLE
Supporting h264/WebRTC for separate capture devices?

### ISSUE THREAD

**Issue by aggieNick02** [pikvm/kvmd#52]
I was interested in working on and contributing a PR to allow the new WebRTC functionality to be used in Pi-KVM when the video capture and encoding is occurring on another machine. If this sounds ok, just let me know. You could either assign the issue to me or close it and I'll come back when I have the PR ready. Here are the details:

I use Pi-KVM in a couple ways. The first is the fairly standard setup with capture via CSI device. The second is a bit different. I use the Pi and kvmd to do the keyboard and mouse control, but have a custom capture server running on a different machine (allows for capture at higher resolutions and framerates than CSI can do).

For mjpeg, all I had to do was override the streamer config to:
```
    streamer:
        unix: ''
        host: {{ capture_host_ip }}
        port: {{ capture_host_ustreamer_port }}
        cmd:
            - "/bin/sleep"
            - "infinity"
```

and then make sure my capture host implemented the same web API as ustreamer. The modularity of your system is nice and being able to have capture on one machine and keyboard/mouse on another could provide some nice upgrade paths in the future where the device doing capture would not need to also have USB OTG.

Anyways, with the new h264 webrtc support, doing this isn't as easy. I was able to prototype getting it working... the basic steps are:

1. Build janus with the streaming plugin enabled (https://janus.conf.meetecho.com/docs/streaming.html)
2. Replace the jcfg file for ustreamer with a basic jcfg file for a simple janus streamer that takes an rtp stream on a port
3. Modify stream.js to use the janus streaming plugin (and associated id) instead of the janus ustreamer plugin
4. Have the capture device transmit an rtp stream to the given port on the kvmd/janus device

Only step 3 seems a bit hacky. I"m imagining the name of the janus plugin and other data needed could be placed in the config file with defaults for the normal ustreamer case, and then overriding would cause stream.js to behave differently as needed. Thoughts?

**Comment by mdevaev** [comment]
I would prefer that you make something like a small daemon for the ustreamer plugin to feed it a video stream. Thus, you will only need to replace the ustreamer startup command with your daemon, and you will not have to rewrite anything else. This will be easy enough to do if you use my C code to work with shared memory.

**Comment by aggieNick02** [comment]
Ah, I think I understand. So the ustreamer janus plugin would stay as is, and my daemon would fill the memory sink instead of ustreamer writing to the sink?

I think I can do that with my current case pretty easily. Had a couple questions to try to understand better the benefits of going through the ustreamer plugin.

Are there other benefits besides minimal code modification to passing the frames through the ustreamer plugin vs the janus streaming plugin? There are actually some devices out there though that do all the encoding and rtp wrapping in hardware and ship the result to an rtp endpoint. (Examples include cards like the [Matrox Mura IPX encode line](https://www.matrox.com/en/video/products/video-walls/mura-ipx-series/4k-capture-ip-encode-decode) that acutally have a network port on it.) In that case the pi daemon would have to unwrap the rtp to get the raw frames to hand off to the ustreamer plugin, which would then wrap them back up.

Even in my case where my capture machine is holding the encoded frames in memory, it may make sense to wrap them in rtp for transmission to the pi-kvm (versus shipping the frames with a simple custom protocol). If I did use rtp to ship to the Pi-KVM, It  feels a little weird to have a small daemon whose job is to receive and parse an rtp stream to get the frames to hand to the ustreamer plugin to build back up a new rtp stream.

Thanks for helping me understand the best way to accomplish this.

**Comment by mdevaev** [comment]
It seems to me that the minimal footprint is quite a weighty argument, especially considering that the janus streaming plugin is a huge harvester for many thousands of lines. By using it, you will first increase the entropy inside the kvmd code to support it. Secondly, you will have to deal with its bugs, while the ustreamer plugin is perfectly tested.

As for RTP, the repackaging operation is easy enough to be ignored. For example, for gstreamer, this is quite typical for many pipelines. And if you make a custom protocol (for example, if you transmit frames over tcp, and then put them in shared memory), then everything will be even easier.

**Comment by aggieNick02** [comment]
Those are both good and fair points. I turned to the streaming plugin and sending RTP after thinking a little bit about the shared memory approach just because it was quicker for me to implement - I'd read the shared memory code but hadn't followed it 100%. I had a quick question about the shared memory approach to make sure I do it right from the server side.

The shared memory holds space for a single frame. Is there anything that stops the server (ustreamer) from writing faster than the client (ustreamer plugin) reads, which could lead to frames being missed on the plugin side? I didn't see anything like that present - perhaps with the capture and encoding happening on the pi, it just doesn't ever happen.

But I could see the timing characteristics being different when frames are coming over the network, with multiple video frames arriving in a burst and my little daemon needing to do something to make sure a frame, especially a key frame, is not overwritten in shared memory before it is read by the janus ustreamer plugin. Does that make sense? Sorry if there is protection against this or a reason it isn't a problem and I missed it.

**Comment by mdevaev** [comment]
Indeed, there is no protection against this and nothing can be done here without organizing a full-fledged queue. I just made the polling interval small enough that it wouldn't cause any problems. Practice shows that this works quite well.

**Comment by aggieNick02** [comment]
Sounds good, thank you.

**Comment by mdevaev** [comment]
I advise you to build your application based on the ustreamer's libs like logging, memsink, etc, then you can reuse a significant part of the code.
